session-state-machine/src/main/scala/zio/raft/sessionstatemachine/SessionStateMachine.scala

```diff
@@ -0,0 +1,573 @@
+    * iterating through all pending requests.
+    */
+  private def handleGetRequestsForRetry(cmd: SessionCommand.GetRequestsForRetry[SR])
+    : State[HMap[Schema], List[PendingServerRequest[SR]]] =
+    State.modify { state =>
+      // Use foldRight to collect updated requests and build new state
+      // This avoids vars and the need to reverse at the end
+      state.iterator["serverRequests"].foldRight((List.empty[PendingServerRequest[SR]], state)) {
+        case (((sessionId, requestId), pendingAny), (accumulated, currentState)) =>
+          val pending = pendingAny.asInstanceOf[PendingServerRequest[SR]]
+
+          // Check if this request needs retry (lastSentAt before threshold)
+          if pending.lastSentAt.isBefore(cmd.lastSentBefore) then
+            // Update lastSentAt and add to accumulated list
+            val updatedReq = pending.copy(lastSentAt = cmd.createdAt)
+            val updatedState = currentState.updated["serverRequests"]((sessionId, requestId), updatedReq)
+            (updatedReq :: accumulated, updatedState)
+          else
+            (accumulated, currentState)
+      }
+    }
+
+  // ====================================================================================
+  // SERVER REQUEST MANAGEMENT
+  // ====================================================================================
+
+  /** Add server requests to state and assign monotonically increasing IDs.
+    *
+    * Each request is stored with a composite key (SessionId, RequestId) for efficiency. sessionId comes from
+    * ServerRequestForSession, not as a parameter!
+    *
+    * @param createdAt
+    *   Timestamp when the command was created (for lastSentAt)
+    * @param serverRequests
+    *   Chunk of server requests (each contains target sessionId)
+    */
+  private def addServerRequests(
+    createdAt: Instant,
+    serverRequests: Chunk[ServerRequestForSession[SR]]
+  ): State[HMap[Schema], List[ServerRequestWithContext[SR]]] =
+    if serverRequests.isEmpty then
+      State.succeed(Nil)
+    else
+      State.modify { state =>
+        // Group requests by sessionId for efficient ID assignment
+        val groupedBySession = serverRequests.groupBy(_.sessionId)
+
+        // Process each session's requests
+        val (finalState, allRequestsWithContext) =
+          groupedBySession.foldLeft((state, List.empty[ServerRequestWithContext[SR]])) {
+            case ((currentState, accumulated), (sessionId, sessionRequests)) =>
+              // Get last assigned ID for this session
+              val lastId = currentState.get["lastServerRequestId"](sessionId)
+                .getOrElse(RequestId.zero)
+
+              // Assign IDs starting from lastId + 1
+              val requestsWithIds = sessionRequests.zipWithIndex.map { case (reqWithSession, index) =>
+                val newId = RequestId(RequestId.unwrap(lastId) + index + 1)
```

Comment:
I would like to do it without unwrap, first we can use lastId.next, and then, let's add extension method to RequestId that allow us to add integer.